include/dlaf/eigensolver/tridiag_solver/impl.h

```diff
@@ -188,51 +190,76 @@ void offloadDiagonal(Matrix<const T, Device::CPU>& tridiag, Matrix<T, D>& evals)
+                     Matrix<T, D>(vec_size, vec_tile_size),          // z0
+                     Matrix<T, D>(vec_size, vec_tile_size),          // z1
                      Matrix<SizeType, D>(vec_size, vec_tile_size)};  // i2
 
-  WorkSpaceHost<T> ws_h{Matrix<T, Device::CPU>(vec_size, vec_tile_size),         // dtmp
-                        Matrix<SizeType, Device::CPU>(vec_size, vec_tile_size),  // i1
-                        Matrix<SizeType, Device::CPU>(vec_size, vec_tile_size),  // i3
-                        Matrix<ColType, Device::CPU>(vec_size, vec_tile_size)};  // c
+  WorkSpaceHost<T> ws_h{Matrix<T, Device::CPU>(vec_size, vec_tile_size),          // d0
+                        Matrix<ColType, Device::CPU>(vec_size, vec_tile_size),    // c
+                        Matrix<SizeType, Device::CPU>(vec_size, vec_tile_size),   // i1
+                        Matrix<SizeType, Device::CPU>(vec_size, vec_tile_size)};  // i3
 
   // Mirror workspace on host memory for CPU-only kernels
-  WorkSpaceHostMirror<T, D> ws_hm{initMirrorMatrix(evals), initMirrorMatrix(ws.mat1),
-                                  initMirrorMatrix(ws.z), initMirrorMatrix(ws.ztmp),
+  WorkSpaceHostMirror<T, D> ws_hm{initMirrorMatrix(ws.e2), initMirrorMatrix(ws.d1),
+                                  initMirrorMatrix(ws.z0), initMirrorMatrix(ws.z1),
                                   initMirrorMatrix(ws.i2)};
 
-  // Set `evecs` to `zero` (needed for Given's rotation to make sure no random values are picked up)
-  matrix::util::set0<B, T, D>(pika::execution::thread_priority::normal, evecs);
+  // Set `ws.e0` to `zero` (needed for Given's rotation to make sure no random values are picked up)
+  matrix::util::set0<B, T, D>(pika::execution::thread_priority::normal, ws.e0);
 
   // Cuppen's decomposition
   auto offdiag_vals = cuppensDecomposition(tridiag);
 
   // Solve with stedc for each tile of `tridiag` (nb x 2) and save eigenvectors in diagonal tiles of
   // `evecs` (nb x nb)
   if constexpr (D == Device::CPU) {
-    solveLeaf(tridiag, evecs);
+    solveLeaf(tridiag, ws.e0);
   }
   else {
-    solveLeaf(tridiag, evecs, ws_hm.mat1);
+    solveLeaf(tridiag, ws.e0, ws_hm.e2);
   }
 
-  // Offload the diagonal from `tridiag` to `evals`
-  offloadDiagonal(tridiag, ws_hm.evals);
+  // Offload the diagonal from `tridiag` to `d0`
+  offloadDiagonal(tridiag, ws_h.d0);
 
   // Each triad represents two subproblems to be merged
   for (auto [i_begin, i_split, i_end] : generateSubproblemIndices(distr.nrTiles().rows())) {
-    mergeSubproblems<B>(i_begin, i_split, i_end, offdiag_vals[to_sizet(i_split - 1)], ws, ws_h, ws_hm,
-                        evals, evecs);
+    mergeSubproblems<B>(i_begin, i_split, i_end, offdiag_vals[to_sizet(i_split - 1)], ws, ws_h, ws_hm);
   }
 
-  copy({0, 0}, evals.distribution().localNrTiles(), ws_hm.evals, evals);
+  const SizeType n = evecs.nrTiles().rows();
+  copy({0, 0}, ws.i2.distribution().localNrTiles(), ws_hm.i2, ws.i2);
```

Comment:
I'm missing something: why do we have to specify begin/size? aren't they distributed the same way?

```suggestion
  copy(ws_hm.i2, ws.i2);
```

There are different places where this is done, for which the question might be similar.